internal/nimparser/v2/nimparser.go

```diff
@@ -0,0 +1,234 @@
+			continue
+		}
+
+		// Determine backend type
+		backend := profile.Tags["llm_engine"]
+		if backend == "" {
+			backend = profile.Tags["backend"]
+		}
+
+		if modelSpec.Engine != "" && !strings.Contains(backend, strings.TrimSuffix(modelSpec.Engine, "_llm")) {
+			continue
+		}
+
+		// Perform GPU match only when optimized engine is selected or GPU filters are provided
+		if isOptimizedEngine(modelSpec.Engine) || len(modelSpec.GPUs) > 0 {
+			// Skip non optimized profiles
+			if !isOptimizedEngine(backend) {
+				continue
+			}
+			if len(modelSpec.GPUs) > 0 || len(discoveredGPUs) > 0 {
+				if !matchGPUProfile(modelSpec, profile, discoveredGPUs) {
+					continue
+				}
+			}
+		}
+
+		// Profile matched the given model parameters, add hash to the selected profiles
+		selectedProfiles = append(selectedProfiles, profile.ID)
+	}
+
+	return selectedProfiles, nil
+}
+
+func (manifest NIMManifest) GetProfilesList() []string {
+
+	profileIDs := make([]string, len(manifest.Profiles))
+
+	for k, profile := range manifest.Profiles {
+		profileIDs[k] = profile.ID
+	}
+	return profileIDs
+}
+func (manifest NIMManifest) GetProfileModel(profileID string) string {
+	return ""
+}
+func (manifest NIMManifest) GetProfileTags(profileID string) map[string]string {
+	for _, profile := range manifest.Profiles {
+		if profileID == profile.ID {
+			return profile.Tags
+		}
+	}
+	return nil
+}
+func (manifest NIMManifest) GetProfileRelease(profileID string) string {
+	return ""
+}
+
+func isOptimizedEngine(engine string) bool {
+	return engine != "" && strings.Contains(strings.ToLower(engine), BackendTypeTensorRT)
+}
+
+func matchGPUProfile(modelSpec appsv1alpha1.ModelSpec, profile NIMProfile, discoveredGPUs []string) bool {
+	foundGPU := false
+
+	for _, gpu := range modelSpec.GPUs {
+		// Check for GPU product match
+		if gpu.Product != "" {
+			// Check if the product matches the "gpu" tag
+			if strings.Contains(strings.ToLower(profile.Tags["gpu"]), strings.ToLower(gpu.Product)) {
+				foundGPU = true
+			}
+
+			// Check if the product matches the "key" tag
+			if strings.Contains(strings.ToLower(profile.Tags["key"]), strings.ToLower(gpu.Product)) {
+				foundGPU = true
+			}
+
+			// If the GPU product matches, check the GPU IDs
+			if foundGPU && len(gpu.IDs) > 0 {
+				foundID := false
+				for _, id := range gpu.IDs {
+					if id == strings.TrimSuffix(profile.Tags["gpu_device"], ":10de") {
+						foundID = true
+						break
+					}
+				}
+
+				// If the GPU product matches but no IDs match, return false
+				if !foundID {
+					return false
+				}
+			}
+		}
+	}
+
+	// If a GPU product was matched and IDs (if any) also matched, return true
+	if foundGPU {
+		return true
+	}
+
+	// If no match was found in the specified GPUs, check the discovered GPUs
+	for _, productLabel := range discoveredGPUs {
+		if productLabel != "" {
+			// match for llm nim format
+			if strings.Contains(strings.ToLower(productLabel), strings.ToLower(profile.Tags["gpu"])) {
```

Comment:
here we need to change to check for buildable profile tag, if set to true, then we need to select this profile.

Reply:
Fixed

Reply:
looks good.